From ee4faa1980f6001336cb20b09a9fa6fb3ac221ab Mon Sep 17 00:00:00 2001
From: Shengyu Qu <wiagn@4d2.org>
Date: Mon, 18 May 2026 19:23:00 +0800
Subject: [PATCH] Revert "renderer: never use linear downscaling on HDR
 sources"

This reverts commit 030b59d1771cdfe7c563480aa187dbc11645ca36.

Since we have antiringing support for EWA scalers now, we have no need to
not use linear scaling for HDR content.

Closes: https://github.com/haasn/libplacebo/issues/281

Signed-off-by: Shengyu Qu <wiagn@4d2.org>
---
 src/renderer.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/renderer.c b/src/renderer.c
index ec04c3be..83e58864 100644
--- a/src/renderer.c
+++ b/src/renderer.c
@@ -2031,10 +2031,12 @@ static bool pass_scale_main(struct pass_state *pass)
     if (params->disable_linear_scaling || fbofmt->component_depth[0] < 16)
         use_sigmoid = use_linear = false;
 
-    // Avoid sigmoidization for HDR content because it clips to [0,1], and
-    // linearization because it causes very nasty ringing artefacts.
-    if (pl_color_space_is_hdr(&img->color))
-        use_sigmoid = use_linear = false;
+    // Avoid sigmoidization for HDR content because it clips to [0,1]
+    if (pl_color_space_is_hdr(&img->color)) {
+        use_sigmoid = false;
+        if (fbofmt->type != PL_FMT_FLOAT)
+            use_linear = false; // linear HDR needs out-of-range signals
+    }
 
     if (!(use_linear || use_sigmoid) && img->color.transfer == PL_COLOR_TRC_LINEAR) {
         img->color.transfer = image->color.transfer;
-- 
GitLab

